doc/plugin_tutorial/tuto2/src/persistent_counter.ml

```diff
@@ -0,0 +1,57 @@
+(*
+ * This file defines our persistent counter, which we use in the
+ * CountPersistent command.
+ *)
+
+(*
+ * As before, we want to override references:
+ *)
+open Summary.Local
+
+(*
+ * At its core, our persistent counter looks exactly the same as
+ * our non-persistent counter (with a different name to prevent collisions):
+ *)
+let counter = ref ~name:"persistent_counter" 0
+
+(*
+ * The difference is that we need to declare it as a persistent object
+ * using Libobject.declare_object. To do that, we define a function that
+ * saves the value that is passed to it into the reference we have just defined:
+ *)
+let cache_count (_, v) =
+  counter := v
+
+(*
+ * We then use declare_object to create a function that takes an integer value
+ * and creates a persistent object from that value:
+ *)
+let declare_counter : int -> Libobject.obj =
+  let open Libobject in
+  declare_object
+    {
+      (default_object "COUNTER") with
+      cache_function = cache_count;
+      load_function = (fun _ -> cache_count);
+      open_function = (fun _ -> cache_count);
+    }
```

Comment:
Is this the intended module semantics?
~~~coq
Module M.
Count Persistent. (* 1 *)
End M.
Count Persistent. (* 2 *)
Count Persistent. (* 3 *)
Import M.
Count Persistent. (* 2 *)
Count Persistent. (* 3 *)
Import M.
Count Persistent. (* 2 *)
Count Persistent. (* 3 *)
~~~
(Require acts like End btw)

Reply:
I hadn't actually thought of that distinction. I'm not sure what I would want it to do in that case. 

Which of the hooks would you change if you wanted it to not reset every time you Import M above? I supposed if it involves two hooks that act differently, it's probably better to show that anyways for the sake of demonstrating the different hooks.

Reply:
I supposed the open function would just do nothing?

Reply:
open is Import, load is End/Require
btw I think we want to discourage (not very strongly though) require-time effects as they're hard to control / non modular.

also note that with this design if you require something which had an increment it will override, ie:
a.v:
~~~coq
Count Persistent. (* 1 *)
~~~
b.v:
~~~coq
Count Persistent. (* 1 *)
Count Persistent. (* 2 *)
Require a.
Count Persistent. (* 2 *)
Require a.
Count Persistent. (* 3: redoing require has no effect IIRC *)
~~~
(well maybe not because the default classify_function is Keep and that has weird semantics, but we should really change that at some point to force using an explicit classify_function IMO)

I would rather have done something like
~~~ocaml 
let counter = Summary.ref ~name:"tuto-counter" 0

let act (_,()) = incr counter 

(* It's convention to call these functions in_foo because they cast the specific type of the object to the generic obj. With declare_object_full you also get a out_foo in the other direction. *)
let in_counter : unit -> Libobject.obj =
  let open Libobject in
  declare_object
    {
      (default_object "COUNTER") with
      cache_function = act;
      open_function = (fun _ () -> act ()); (* use load_function instead for require-time effect *)
    }

let increment () = Lib.add_anonymous_leaf (in_counter ())
~~~

In general the stuff in a libobject should look like a diff for your state (here it's trivial so it's unit), not a snapshot of the state.